File: app/clients/openrouter.py
import json
from time import perf_counter
from typing import Any

import httpx

from app.config import Settings
from app.utils import extract_urls, safe_json_loads, unique_preserve_order


class OpenRouterClient:
    def __init__(self, settings: Settings):
        self.settings = settings

    @property
    def is_configured(self) -> bool:
        return bool(self.settings.openrouter_api_key)

    def chat(self, model: str, system_prompt: str, user_prompt: str, temperature: float = 0.2, max_tokens: int = 4000) -> str:
        data, _ = self._chat_completion(
            model=model,
            system_prompt=system_prompt,
            user_prompt=user_prompt,
            temperature=temperature,
            max_tokens=max_tokens,
            enable_web_search=False,
        )
        return self._extract_message_text(data)

    def execute_prompt(
        self,
        model: str,
        system_prompt: str,
        user_prompt: str,
        temperature: float = 0,
        max_tokens: int = 1200,
        enable_web_search: bool = False,
    ) -> tuple[str, list[str], int]:
        data, latency_ms = self._chat_completion(
            model=model,
            system_prompt=system_prompt,
            user_prompt=user_prompt,
            temperature=temperature,
            max_tokens=max_tokens,
            enable_web_search=enable_web_search,
        )
        text = self._extract_message_text(data)
        urls = unique_preserve_order(extract_urls(text) + self._extract_urls(data))
        return text, urls, latency_ms

    def _chat_completion(
        self,
        model: str,
        system_prompt: str,
        user_prompt: str,
        temperature: float,
        max_tokens: int,
        enable_web_search: bool,
    ) -> tuple[dict[str, Any], int]:
        if not self.is_configured:
            raise RuntimeError("OPENROUTER_API_KEY is not configured")

        payload = {
            "model": model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            "temperature": temperature,
            "max_tokens": max_tokens,
        }
        if enable_web_search and self.settings.openrouter_web_plugin_id:
            payload["plugins"] = [{"id": self.settings.openrouter_web_plugin_id}]

        start = perf_counter()
        with httpx.Client(timeout=self.settings.request_timeout_seconds) as client:
            response = client.post(f"{self.settings.openrouter_base_url}/chat/completions", headers=self._headers(), json=payload)
            response.raise_for_status()
            data = response.json()
        latency_ms = int((perf_counter() - start) * 1000)
        return data, latency_ms

    def _headers(self) -> dict[str, str]:
        headers = {
            "Authorization": f"Bearer {self.settings.openrouter_api_key}",
            "Content-Type": "application/json",
            "X-Title": self.settings.openrouter_app_name,
        }
        if self.settings.openrouter_site_url:
            headers["HTTP-Referer"] = self.settings.openrouter_site_url
        return headers

    def _extract_message_text(self, data: dict[str, Any]) -> str:
        message = data["choices"][0]["message"]["content"]
        if isinstance(message, list):
            parts = []
            for chunk in message:
                if isinstance(chunk, dict) and chunk.get("type") == "text":
                    parts.append(chunk.get("text", ""))
            return "\n".join(part for part in parts if part)
        return str(message)

    def _extract_urls(self, data: Any) -> list[str]:
        urls: list[str] = []

        def walk(node: Any) -> None:
            if isinstance(node, dict):
                for key, value in node.items():
                    if key in {"url", "uri"} and isinstance(value, str):
                        urls.append(value)
                    walk(value)
            elif isinstance(node, list):
                for value in node:
                    walk(value)

        walk(data)
        return urls

    def chat_json(self, model: str, system_prompt: str, user_prompt: str, temperature: float = 0.2, max_tokens: int = 4000) -> dict:
        raw = self.chat(model=model, system_prompt=system_prompt, user_prompt=user_prompt, temperature=temperature, max_tokens=max_tokens)
        parsed = safe_json_loads(raw)
        if not isinstance(parsed, dict):
            raise json.JSONDecodeError("Expected JSON object", raw, 0)
        return parsed

File: src/lib/openrouter.ts
import { env, assertOpenRouterKey } from "@/lib/env";
import { extractUrls, safeJsonParse, uniquePreserveOrder } from "@/lib/utils";

interface ChatOptions {
  model: string;
  systemPrompt: string;
  userPrompt: string;
  requestId?: string;
  temperature?: number;
  maxTokens?: number;
  enableWebSearch?: boolean;
}

interface OpenRouterMessage {
  content: string | Array<{ type?: string; text?: string }>;
}

interface OpenRouterResponse {
  choices?: Array<{ message?: OpenRouterMessage }>;
}

export async function openRouterChat(options: ChatOptions): Promise<string> {
  const { data } = await requestChatCompletion(options);
  return extractMessageText(data);
}

export async function openRouterChatJson<T>(options: ChatOptions): Promise<T> {
  const raw = await openRouterChat(options);
  return safeJsonParse<T>(raw);
}

export async function openRouterExecutePrompt(options: ChatOptions): Promise<{ text: string; urls: string[]; latencyMs: number }> {
  const { data, latencyMs } = await requestChatCompletion(options);
  const text = extractMessageText(data);
  const urls = uniquePreserveOrder([...extractUrls(text), ...extractNestedUrls(data)]);
  return { text, urls, latencyMs };
}

async function requestChatCompletion({
  model,
  systemPrompt,
  userPrompt,
  requestId,
  temperature = 0.2,
  maxTokens = 4000,
  enableWebSearch = false,
}: ChatOptions): Promise<{ data: OpenRouterResponse; latencyMs: number }> {
  const apiKey = assertOpenRouterKey();
  const timeoutSeconds = Math.max(5, Number.isFinite(env.requestTimeoutSeconds) ? env.requestTimeoutSeconds : 60);
  const timeoutMs = timeoutSeconds * 1000;

  const payload: Record<string, unknown> = {
    model,
    messages: [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ],
    temperature,
    max_tokens: maxTokens,
  };

  if (enableWebSearch && env.openRouterWebPluginId) {
    payload.plugins = [{ id: env.openRouterWebPluginId }];
  }

  const startedAt = Date.now();
  const timeoutController = new AbortController();
  const timeoutHandle = setTimeout(() => timeoutController.abort(), timeoutMs);

  let response: Response;
  try {
    response = await fetch(`${env.openRouterBaseUrl}/chat/completions`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
        ...(requestId ? { "X-Request-Id": requestId } : {}),
        ...(env.openRouterSiteUrl ? { "HTTP-Referer": env.openRouterSiteUrl } : {}),
        "X-Title": env.openRouterAppName,
      },
      body: JSON.stringify(payload),
      cache: "no-store",
      signal: timeoutController.signal,
    });
  } catch (error) {
    if (timeoutController.signal.aborted) {
      throw new Error(`OpenRouter timeout after ${timeoutSeconds}s for model ${model}`);
    }
    throw error;
  } finally {
    clearTimeout(timeoutHandle);
  }

  if (!response.ok) {
    const detail = await response.text();
    throw new Error(`OpenRouter error ${response.status}: ${detail}`);
  }

  const data = (await response.json()) as OpenRouterResponse;
  return { data, latencyMs: Date.now() - startedAt };
}

function extractMessageText(data: OpenRouterResponse): string {
  const message = data.choices?.[0]?.message?.content;
  if (!message) {
    return "";
  }
  if (typeof message === "string") {
    return message;
  }
  return message
    .map((chunk) => (typeof chunk.text === "string" ? chunk.text : ""))
    .filter(Boolean)
    .join("\n");
}

function extractNestedUrls(value: unknown): string[] {
  const urls: string[] = [];

  const walk = (node: unknown) => {
    if (Array.isArray(node)) {
      node.forEach(walk);
      return;
    }
    if (!node || typeof node !== "object") {
      return;
    }

    for (const [key, nestedValue] of Object.entries(node)) {
      if ((key === "url" || key === "uri") && typeof nestedValue === "string") {
        urls.push(nestedValue);
      }
      walk(nestedValue);
    }
  };

  walk(value);
  return urls;
}
